File: app/src/main/kotlin/tech/mmarca/openvitals/domain/model/SleepModels.kt
package tech.mmarca.openvitals.domain.model

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset

data class SleepData(
    val id: String,
    val startTime: Instant,
    val endTime: Instant,
    val durationMs: Long,
    val source: String,
    val title: String? = null,
    val notes: String? = null,
    val startZoneOffset: ZoneOffset? = null,
    val endZoneOffset: ZoneOffset? = null,
    val lastModifiedTime: Instant? = null,
    val clientRecordId: String? = null,
    val clientRecordVersion: Long? = null,
    val recordingMethod: Int? = null,
    val device: SleepDeviceData? = null,
    val stages: List<SleepStage> = emptyList(),
) {
    val durationHours: Double get() = durationMs / 3_600_000.0
}

data class SleepDeviceData(
    val type: Int,
    val manufacturer: String?,
    val model: String?,
)

data class SleepStage(
    val startTime: Instant,
    val endTime: Instant,
    val stageType: Int,
) {
    val durationMs: Long get() = endTime.toEpochMilli() - startTime.toEpochMilli()

    companion object {
        const val STAGE_AWAKE = 1
        const val STAGE_SLEEPING = 2
        const val STAGE_OUT_OF_BED = 3
        const val STAGE_LIGHT = 4
        const val STAGE_DEEP = 5
        const val STAGE_REM = 6
        const val STAGE_AWAKE_IN_BED = 7
    }
}

/**
 * Time actually ASLEEP: stage durations with the awake epochs excluded. A
 * night with 8h in bed and 40 minutes awake reads 7h20m. Sessions whose
 * writer recorded no stages keep the plain session duration.
 */
fun SleepData.asleepDurationMs(): Long = sleepDurationMsFromStages(stages, durationMs)

data class DailySleepDuration(
    val date: LocalDate,
    val durationMs: Long,
) {
    val durationHours: Double get() = durationMs / 3_600_000.0
}

data class SleepReadData(
    val sessions: List<SleepData> = emptyList(),
    val dailyAggregateDurations: List<DailySleepDuration> = emptyList(),
)

/**
 * The union of slept intervals in milliseconds — a sweep-merge, so overlapping
 * cross-source survivors can never sum past real time.
 */
fun sleepSessionsUnionMs(sessions: Iterable<SleepData>): Long {
    val intervals = sessions
        .filter { it.endTime.isAfter(it.startTime) }
        .map { it.startTime to it.endTime }
        .sortedBy { it.first }
    if (intervals.isEmpty()) return 0L
    var totalMs = 0L
    var currentStart = intervals.first().first
    var currentEnd = intervals.first().second
    for ((start, end) in intervals.drop(1)) {
        if (start.isAfter(currentEnd)) {
            totalMs += currentEnd.toEpochMilli() - currentStart.toEpochMilli()
            currentStart = start
            currentEnd = end
        } else if (end.isAfter(currentEnd)) {
            currentEnd = end
        }
    }
    return totalMs + (currentEnd.toEpochMilli() - currentStart.toEpochMilli())
}

/**
 * Concatenates the stages of [orderedSessions], deduplicating identical
 * (start, end, type) triples, then fills each inter-segment gap no wider than
 * [maxGap] with an explicit Awake stage — a wake-split night keeps continuous
 * stage coverage and its schedule bar has no hole.
 */
fun combineNightStages(
    orderedSessions: List<SleepData>,
    maxGap: java.time.Duration,
): List<SleepStage> {
    val seen = mutableSetOf<Triple<java.time.Instant, java.time.Instant, Int>>()
    val stages = mutableListOf<SleepStage>()
    for (stage in orderedSessions.flatMap { it.stages }) {
        if (seen.add(Triple(stage.startTime, stage.endTime, stage.stageType))) {
            stages += stage
        }
    }
    if (stages.isEmpty()) return emptyList()
    // A gap between two sessions is time OUT of bed, not time awake in it. Typing it
    // STAGE_AWAKE made a 90-minute get-up read as wake-after-sleep-onset: it inflated the
    // Awake row, and through it sleep efficiency and the score's WASO term. STAGE_OUT_OF_BED
    // is excluded from AwakeStageTypes and from SleepScore.isAwakeStage, so the gap still
    // shows on the timeline without being counted as restless time in bed. Its overlap rank
    // is the lowest of any real stage, so a recorded stage always wins the disputed region.
    val gapStages = orderedSessions.zipWithNext().mapNotNull { (previous, next) ->
        val gap = java.time.Duration.between(previous.endTime, next.startTime)
        if (!gap.isNegative && gap > java.time.Duration.ZERO && gap <= maxGap) {
            SleepStage(previous.endTime, next.startTime, SleepStage.STAGE_OUT_OF_BED)
        } else {
            null
        }
    }
    return resolveSleepStages(
        stages = (stages + gapStages)
            .sortedWith(compareBy<SleepStage> { it.startTime }.thenBy { it.endTime }),
        sessionStart = orderedSessions.minOf { it.startTime },
        sessionEnd = orderedSessions.maxOf { it.endTime },
    )
}

/** Sum of stage durations, negative stages ignored. */
fun List<SleepStage>.totalStageMs(): Long = sumOf { it.durationMs.coerceAtLeast(0L) }

/** A fully-staged night is ~1.0; below this the hypnogram would mostly be a guess. */
const val MinSleepStageCoverage = 0.5

/**
 * Whether the session's stages cover enough of its span to draw a truthful
 * hypnogram. Coverage = summed stage time over the full session span.
 */
fun sleepSessionHasReliableStages(
    session: SleepData,
    minCoverage: Double = MinSleepStageCoverage,
): Boolean {
    if (session.stages.isEmpty()) return false
    val spanMs = session.endTime.toEpochMilli() - session.startTime.toEpochMilli()
    if (spanMs <= 0L) return false
    return session.stages.totalStageMs().toDouble() / spanMs >= minCoverage
}

/** Stage types shown as "Awake" in grouped breakdowns (in-bed awake, excluding out-of-bed). */
val AwakeStageTypes: Set<Int> = setOf(SleepStage.STAGE_AWAKE, SleepStage.STAGE_AWAKE_IN_BED)

/** Stage types shown as "Core" (Apple naming) — light plus generic sleeping. */
val CoreStageTypes: Set<Int> = setOf(SleepStage.STAGE_LIGHT, SleepStage.STAGE_SLEEPING)

/** Total duration of the stages whose [SleepStage.stageType] is in [types], overlaps counted once. */
fun List<SleepStage>.durationMsForTypes(types: Set<Int>): Long =
    filter { it.stageType in types }.unionDurationMs()

internal fun sleepDurationMsFromStages(
    stages: List<SleepStage>,
    fallbackDurationMs: Long,
): Long {
    if (stages.isEmpty()) return fallbackDurationMs.coerceAtLeast(0L)

    // Union, not a sum: a writer recording the same stretch under two stages
    // must never make asleep time exceed time in bed.
    val sleepStageDurationMs = stages
        .filter { it.stageType.isSleepDurationStage() }
        .unionDurationMs()

    return sleepStageDurationMs.takeIf { it > 0L } ?: fallbackDurationMs.coerceAtLeast(0L)
}

private fun Int.isSleepDurationStage(): Boolean = when (this) {
    SleepStage.STAGE_SLEEPING,
    SleepStage.STAGE_LIGHT,
    SleepStage.STAGE_DEEP,
    SleepStage.STAGE_REM -> true
    else -> false
}

/**
 * When a writer records overlapping stage intervals for one session — the
 * bug shape: one stretch stored as BOTH light and deep — every instant must
 * belong to exactly one stage. This resolves the overlaps the way Google Fit
 * renders the same data: the deeper stage wins the disputed region.
 */
private fun sleepStageOverlapRank(stageType: Int): Int = when (stageType) {
    SleepStage.STAGE_DEEP -> 7
    SleepStage.STAGE_REM -> 6
    SleepStage.STAGE_LIGHT -> 5
    SleepStage.STAGE_SLEEPING -> 4
    SleepStage.STAGE_AWAKE_IN_BED -> 3
    SleepStage.STAGE_AWAKE -> 2
    SleepStage.STAGE_OUT_OF_BED -> 1
    else -> 0
}

/**
 * The canonical stage timeline for a session: stages clipped to
 * [sessionStart, sessionEnd], overlaps resolved by [sleepStageOverlapRank],
 * adjacent fragments of one winner merged. All stored [SleepData.stages] pass
 * through here, so every duration, percentage, and hypnogram downstream
 * describes the same non-overlapping intervals. An already-clean list is
 * returned as-is.
 */
fun resolveSleepStages(
    stages: List<SleepStage>,
    sessionStart: Instant,
    sessionEnd: Instant,
): List<SleepStage> {
    if (stages.isEmpty()) return stages
    if (stagesAlreadyResolved(stages, sessionStart, sessionEnd)) return stages

    val clipped = stages.mapNotNull { stage ->
        val start = maxOf(stage.startTime, sessionStart)
        val end = minOf(stage.endTime, sessionEnd)
        if (end.isAfter(start)) stage.copy(startTime = start, endTime = end) else null
    }
    if (clipped.isEmpty()) return emptyList()

    val boundaries = clipped
        .flatMap { listOf(it.startTime, it.endTime) }
        .distinct()
        .sorted()
    val resolved = mutableListOf<SleepStage>()
    boundaries.zipWithNext().forEach { (segmentStart, segmentEnd) ->
        val winner = clipped
            .filter { !it.startTime.isAfter(segmentStart) && !it.endTime.isBefore(segmentEnd) }
            .maxByOrNull { sleepStageOverlapRank(it.stageType) }
            ?: return@forEach // an uncovered gap between stages stays a gap
        val previous = resolved.lastOrNull()
        if (previous != null && previous.stageType == winner.stageType && previous.endTime == segmentStart) {
            resolved[resolved.lastIndex] = previous.copy(endTime = segmentEnd)
        } else {
            resolved += SleepStage(segmentStart, segmentEnd, winner.stageType)
        }
    }
    return resolved
}

private fun stagesAlreadyResolved(
    stages: List<SleepStage>,
    sessionStart: Instant,
    sessionEnd: Instant,
): Boolean {
    var previousEnd = sessionStart
    for (stage in stages) {
        if (stage.startTime.isBefore(previousEnd)) return false
        if (!stage.endTime.isAfter(stage.startTime)) return false
        previousEnd = stage.endTime
    }
    return !previousEnd.isAfter(sessionEnd)
}

private fun List<SleepStage>.unionDurationMs(): Long {
    val intervals = filter { it.endTime.isAfter(it.startTime) }.sortedBy { it.startTime }
    if (intervals.isEmpty()) return 0L
    var totalMs = 0L
    var currentStart = intervals.first().startTime
    var currentEnd = intervals.first().endTime
    for (stage in intervals.drop(1)) {
        if (stage.startTime.isAfter(currentEnd)) {
            totalMs += currentEnd.toEpochMilli() - currentStart.toEpochMilli()
            currentStart = stage.startTime
            currentEnd = stage.endTime
        } else if (stage.endTime.isAfter(currentEnd)) {
            currentEnd = stage.endTime
        }
    }
    return totalMs + (currentEnd.toEpochMilli() - currentStart.toEpochMilli())
}

File: lib/domain/model/sleep_models.dart
import 'dart:math' as math;

import 'package:freezed_annotation/freezed_annotation.dart';

import '../../core/time/local_date.dart';

part 'sleep_models.freezed.dart';

@freezed
abstract class SleepData with _$SleepData {
  const SleepData._();

  const factory SleepData({
    required String id,
    required DateTime startTime,
    required DateTime endTime,
    required int durationMs,
    required String source,
    String? title,
    String? notes,
    Duration? startZoneOffset,
    Duration? endZoneOffset,
    DateTime? lastModifiedTime,
    String? clientRecordId,
    int? clientRecordVersion,
    int? recordingMethod,
    SleepDeviceData? device,
    @Default(<SleepStage>[]) List<SleepStage> stages,
  }) = _SleepData;

  double get durationHours => durationMs / 3600000.0;
}

@freezed
abstract class SleepDeviceData with _$SleepDeviceData {
  const factory SleepDeviceData({
    required int type,
    required String? manufacturer,
    required String? model,
  }) = _SleepDeviceData;
}

@freezed
abstract class SleepStage with _$SleepStage {
  const SleepStage._();

  const factory SleepStage({
    required DateTime startTime,
    required DateTime endTime,
    required int stageType,
  }) = _SleepStage;

  int get durationMs =>
      endTime.millisecondsSinceEpoch - startTime.millisecondsSinceEpoch;

  static const int stageAwake = 1;
  static const int stageSleeping = 2;
  static const int stageOutOfBed = 3;
  static const int stageLight = 4;
  static const int stageDeep = 5;
  static const int stageRem = 6;
  static const int stageAwakeInBed = 7;
}

@freezed
abstract class DailySleepDuration with _$DailySleepDuration {
  const DailySleepDuration._();

  const factory DailySleepDuration({
    required LocalDate date,
    required int durationMs,
  }) = _DailySleepDuration;

  double get durationHours => durationMs / 3600000.0;
}

@freezed
abstract class SleepReadData with _$SleepReadData {
  const factory SleepReadData({
    @Default(<SleepData>[]) List<SleepData> sessions,
    @Default(<DailySleepDuration>[])
    List<DailySleepDuration> dailyAggregateDurations,
  }) = _SleepReadData;
}

/// Stage types shown as "Awake" in grouped breakdowns (in-bed awake, excluding
/// out-of-bed).
const Set<int> awakeStageTypes = {
  SleepStage.stageAwake,
  SleepStage.stageAwakeInBed,
};

/// Stage types shown as "Core" (Apple naming) — light plus generic sleeping.
const Set<int> coreStageTypes = {
  SleepStage.stageLight,
  SleepStage.stageSleeping,
};

extension SleepStageDurations on List<SleepStage> {
  /// Total duration of the stages whose [SleepStage.stageType] is in [types].
  int durationMsForTypes(Set<int> types) => where(
        (stage) => types.contains(stage.stageType),
      ).fold<int>(0, (sum, stage) => sum + math.max(stage.durationMs, 0));

  /// Total milliseconds the stages account for, summed (hypnogram stages do not
  /// overlap). Tells a fully-staged night from one the device only staged part of.
  int get totalStageMs =>
      fold<int>(0, (sum, stage) => sum + math.max(stage.durationMs, 0));
}

/// Below this share of a session's span, its stage data is treated as too partial
/// to draw a hypnogram from. A device that only staged the tail of the night — or
/// a preliminary sync that has not finished — would otherwise render as a
/// near-empty chart with a fragment at one edge. A fully-staged night is ~1.0.
const double kMinSleepStageCoverage = 0.5;

/// Whether [session] has enough stage coverage to draw a meaningful hypnogram:
/// it has stages, and they cover at least [minCoverage] of its span. False for a
/// session the device only partially staged, so callers can show a note instead
/// of a misleading chart.
bool sleepSessionHasReliableStages(
  SleepData session, {
  double minCoverage = kMinSleepStageCoverage,
}) {
  if (session.stages.isEmpty) return false;
  final spanMs = session.endTime.difference(session.startTime).inMilliseconds;
  if (spanMs <= 0) return false;
  return session.stages.totalStageMs / spanMs >= minCoverage;
}

/// Wall-clock milliseconds covered by [sessions], counting overlapping time
/// once. Sweep-merges each session's `[startTime, endTime]` interval, so two
/// overlapping sessions from different sources yield the real time in bed, not
/// the sum of their durations. For non-overlapping input this equals the sum.
int sleepSessionsUnionMs(Iterable<SleepData> sessions) {
  final intervals = [
    for (final session in sessions)
      if (session.endTime.isAfter(session.startTime))
        (session.startTime, session.endTime),
  ]..sort((a, b) => a.$1.compareTo(b.$1));
  if (intervals.isEmpty) return 0;

  var totalMs = 0;
  var currentStart = intervals.first.$1;
  var currentEnd = intervals.first.$2;
  for (final interval in intervals.skip(1)) {
    if (interval.$1.isAfter(currentEnd)) {
      totalMs += currentEnd.difference(currentStart).inMilliseconds;
      currentStart = interval.$1;
      currentEnd = interval.$2;
    } else if (interval.$2.isAfter(currentEnd)) {
      currentEnd = interval.$2;
    }
  }
  totalMs += currentEnd.difference(currentStart).inMilliseconds;
  return totalMs;
}

/// Combines the stages of a night's [orderedSessions] (sorted by start) into one
/// timeline: identical stages are deduped, and the wake gap between consecutive
/// segments (up to [maxGap]) is filled with an Awake stage. This is what lets a
/// night broken by a wake render as one continuous hypnogram — and have the wake
/// count toward the awake share and the reliable-stage coverage — instead of
/// blocks separated by holes. Gaps larger than [maxGap] (a daytime nap, not part
/// of the night) are left unfilled. Empty when no segment carries any stage.
List<SleepStage> combineNightStages(
  List<SleepData> orderedSessions, {
  required Duration maxGap,
}) {
  final seenKeys = <(DateTime, DateTime, int)>{};
  final stages = <SleepStage>[];
  for (final stage in orderedSessions.expand((session) => session.stages)) {
    final key = (stage.startTime, stage.endTime, stage.stageType);
    if (seenKeys.add(key)) stages.add(stage);
  }
  if (stages.isEmpty) return const <SleepStage>[];

  final gapStages = <SleepStage>[];
  for (var index = 0; index < orderedSessions.length - 1; index++) {
    final previous = orderedSessions[index];
    final next = orderedSessions[index + 1];
    final gap = next.startTime.difference(previous.endTime);
    if (!gap.isNegative && gap > Duration.zero && gap <= maxGap) {
      gapStages.add(
        SleepStage(
          startTime: previous.endTime,
          endTime: next.startTime,
          stageType: SleepStage.stageAwake,
        ),
      );
    }
  }

  return [...stages, ...gapStages]..sort((a, b) {
      final byStart = a.startTime.compareTo(b.startTime);
      if (byStart != 0) return byStart;
      return a.endTime.compareTo(b.endTime);
    });
}

int sleepDurationMsFromStages(
  List<SleepStage> stages,
  int fallbackDurationMs,
) {
  if (stages.isEmpty) return math.max(fallbackDurationMs, 0);

  final sleepStageDurationMs = stages
      .where((stage) => _isSleepDurationStage(stage.stageType))
      .fold<int>(0, (sum, stage) => sum + math.max(stage.durationMs, 0));

  return sleepStageDurationMs > 0
      ? sleepStageDurationMs
      : math.max(fallbackDurationMs, 0);
}

bool _isSleepDurationStage(int stageType) {
  switch (stageType) {
    case SleepStage.stageSleeping:
    case SleepStage.stageLight:
    case SleepStage.stageDeep:
    case SleepStage.stageRem:
      return true;
    default:
      return false;
  }
}
